### PR TITLE
clean up media when leaving a meeting

### PR DESCRIPTION
**Pull request by holsted**
# Pull Request Template

## Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Test Coverage

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

**Test Configuration**:
* SDK Version
* Node/Browser Version
* NPM Version

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: packages/node_modules/@webex/plugin-meetings/src/media/index.js [hunk 1]
```diff
@@ -2,10 +2,10 @@
  * Copyright (c) 2015-2018 Cisco Systems, Inc. See LICENSE file.
  */
 /* globals navigator */
-import {
-  AUDIOINPUT,
-  VIDEOINPUT
-} from '../constants';
+
+import Promise from 'bluebird';
```
Comment (chrisadubois) [inline packages/node_modules/@webex/plugin-meetings/src/media/index.js]: yuck but ok

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 2]
```diff
@@ -210,19 +210,33 @@ export default class Meeting extends StatelessSparkPlugin {
     this.mediaId = locus.mediaId;
   }
 
-  setRemoteStreams(pc) {
+  setRemoteStream(pc) {
     // Receiving remote media streams
     pc.ontrack = (event) => {
-      [this.mediaStream, this.localMediaStream] = event.streams;
+      [this.remoteStream] = event.streams;
```
Comment (chrisadubois) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js]: 👍

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 3]
```diff
@@ -210,19 +210,33 @@ export default class Meeting extends StatelessSparkPlugin {
...
 
+      // TODO: It's possible for media to not be present
+      // so we might need to either
+      // A) wait until we have media flowing
+      // B) trigger a second event when video is flowing
       this.trigger('media:ready', {
         type: 'remote',
-        stream: this.mediaStream
+        stream: this.remoteStream
```
Comment (chrisadubois) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js]: hmmmm did all the refs get changed everywhere?

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 4]
```diff
@@ -210,19 +210,33 @@ export default class Meeting extends StatelessSparkPlugin {
...
       });
     };
   }
 
-  setScreenStreams(pc) {
+  unsetRemoteStream() {
+    return Media.stopStream(this.remoteStream).then(() => {
+      this.remoteStream = null;
+
+      this.trigger('media:stopped', {
```
Comment (chrisadubois) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js]: i can fix this, but we should change to the constant EVENT_TRIGGERS.MEDIA_STOPPED

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 5]
```diff
@@ -234,12 +248,22 @@ export default class Meeting extends StatelessSparkPlugin {
...
-      stream: localStream
+      stream: this.localStream
+    });
+  }
+
+  unsetLocalStream() {
+    return Media.stopStream(this.localStream).then(() => {
+      this.localStream = null;
+
+      this.trigger('media:stopped', {
```
Comment (chrisadubois) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js]: same

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 6]
```diff
@@ -249,6 +273,14 @@ export default class Meeting extends StatelessSparkPlugin {
     this.screenPeerConnection = screenPeerConnection;
   }
 
+  unsetPeerConnections() {
+    PeerConnectionManager.close(this.mediaPeerConnection);
```
Comment (chrisadubois) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js]: we should send an event on peerconnections being shut down

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 7]
```diff
@@ -281,7 +314,7 @@ export default class Meeting extends StatelessSparkPlugin {
       })
       .then((streams) => {
         this.state.connect();
-        this.setLocalVideo(streams);
+        this.setLocalStream(streams);
```
Comment (chrisadubois) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js]: 👍

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 8]
```diff
@@ -386,17 +426,31 @@ export default class Meeting extends StatelessSparkPlugin {
...
    */
   leave() {
-    return this.meetingRequest.leaveMeeting({
-      locusUrl: this.locusUrl,
-      selfId: this.selfId,
-      correlationId: this.correlationId,
-      resourceId: null,
-      deviceUrl: this.deviceUrl
-    });
+    return this.meetingRequest
```
Comment (chrisadubois) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js]: we should send an event when the leave completes

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 9]
```diff
@@ -386,17 +426,31 @@ export default class Meeting extends StatelessSparkPlugin {
...
+        resourceId: null,
+        deviceUrl: this.deviceUrl
+      })
+      .then(() => {
+        this.unsetLocalStream();
+        this.unsetRemoteStream();
+        this.unsetPeerConnections();
+      })
+      .then(() => {
+        this.roap.stop(this.correlationId, this.roapSeq);
```
Comment (chrisadubois) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js]: we should send an event when the roap is shut down for a meeting ?

File: packages/node_modules/@webex/plugin-meetings/src/media/index.js (hunk 1 above)
Comment (andrewholsted) [reply packages/node_modules/@webex/plugin-meetings/src/media/index.js]: I could probably remove bluebird and do a different implementation of promise.try if you are passionate about it

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js (hunk 3 above)
Comment (andrewholsted) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js]: ooo. I only changed them in this file. I'll check now

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js (hunk 6 above)
Comment (andrewholsted) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js]: Cool. Will add

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js (hunk 8 above)
Comment (andrewholsted) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js]: Ok. Will add

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js (hunk 9 above)
Comment (andrewholsted) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js]: Sure. Though we need to document where we are sending all these events from so the end users will know what they have available. I'll add a section to the README

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js (hunk 4 above)
Comment (andrewholsted) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js]: Yeah I saw you changed the others and was planning to update after your PR was merged and I rebased on it

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js (hunk 3 above)
Comment (andrewholsted) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js]: I just looked and didn't see anything that depended on mediaStream (outside of the share.js file)

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js (hunk 9 above)
Comment (andrewholsted) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js]: What do you want the event name to be? I was going to move it into roap/handler.js

File: packages/node_modules/@webex/plugin-meetings/src/media/index.js (hunk 1 above)
Comment (chrisadubois) [reply packages/node_modules/@webex/plugin-meetings/src/media/index.js]: we should remove all bluebird at some point

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js (hunk 9 above)
Comment (chrisadubois) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js]: open to ideas, i think we just need placeholders right now.

File: packages/node_modules/@webex/plugin-meetings/src/media/index.js (hunk 1 above)
Comment (andrewholsted) [reply packages/node_modules/@webex/plugin-meetings/src/media/index.js]: blasphemy!

**Pull request merged**